File: 42-trapping-rain-water/42-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int n=height.length;
        int leftMax=height[0];
        int rightMax=height[n-1];
        int ans=0;
        int i=1,j=n-2;
        while(i<=j)
        {
              rightMax=Math.max(rightMax,height[j]);
                leftMax=Math.max(leftMax,height[i]);

            if(leftMax<=rightMax)
            {
                ans+=leftMax-height[i]>0?leftMax-height[i]:0;
                i++;
            }
            else
            {
              ans+=rightMax-height[j]>0?rightMax-height[j]:0;
                j--;  
            }
        }
        return ans;
        
    }
}

File: 42-trapping-rain-water/42-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        
        int n=height.size();
     vector< int >maxl(n);
      vector< int > maxr(n);
        int ans=0;
        if(n==0) return 0;
        maxl[0]=height[0];
        maxr[n-1]=height[n-1];
         for(int j=1;j<n;j++)
            {
                maxl[j]=max(maxl[j-1],height[j]);
            }
         for(int j=n-2;j>=0;j--)
            {
                maxr[j]=max(maxr[j+1],height[j]);
            }
        for(int i=0;i<n;i++)
        {
            ans+=min(maxl[i],maxr[i])-height[i];
        }
       
        return ans;
    }
};